### PR TITLE
Add circleci configs

### PR DESCRIPTION
**Pull request by bqbn**
Once merged, we'll let CircleCI build and push to the environment per every commit.

### CODE REVIEW

**Comment by whimboo** [comment]
@bqbn where does this patch limit the CI from getting triggered on master? We should only run an update job for changes on the raptor branch, but NOT master.

**Comment by bqbn** [comment]
@whimboo, I wasn't aware that we should only deploy from the raptor branch, thus previously this PR doesn't have that limit.

I'm glad that you mentioned it though. I have submitted another commit that limits the build and deploy only to the raptor branch.

We can start testing the deployment once this PR is merged.

Thanks.

File: .utils/deploy.sh
Comment (whimboo) [inline .utils/deploy.sh]: Out of curiosity, how much work will it be in the future to change from `2019` to eg. `2020`? Shall we have a constant at the top for year in use?

File: .utils/deploy.sh
Comment (whimboo) [inline .utils/deploy.sh]: Where does all this code come from? Is it used across projects? Who else from Ops/RelEng would be good to get this reviewed?

File: .circleci/config.yml
Comment (whimboo) [inline .circleci/config.yml]: When is that job run? We should not run it automatically on a fixed schedule, but only when we request a pull manually. That is to not interfere with running performance tests, and resulting different results due to changed media files.

File: .utils/deploy.sh
Comment (bqbn) [reply .utils/deploy.sh]: > Out of curiosity, how much work will it be in the future to change from 2019 to eg. 2020? 

The work involved would be file a PR to change it from "2019" to "2020".

> Shall we have a constant at the top for year in use?

I don't have an opinion on this tbh. Please feel free to update this PR with a constant if you'd like to do so.

File: .utils/deploy.sh
Comment (bqbn) [reply .utils/deploy.sh]: > Where does all this code come from? Is it used across projects?

Yes, I borrowed most of it from some other projects.

> Who else from Ops/RelEng would be good to get this reviewed?

The security headers are mostly secops driven.

@g-k, can you or one of team members help review the headers in this PR and see if they're OK?

To give you some background (and not speaking on behalf of @whimboo ), the site is mainly used by developers and testers, not by public users, although anyone can see it if he/she chooses to. More details in https://bugzilla.mozilla.org/show_bug.cgi?id=1548062 and its related tickets as well.

File: .circleci/config.yml
Comment (bqbn) [reply .circleci/config.yml]: So do you mean that "pulling and installing the media files" should be initiated manually, not when there are commits into the "raptor" branch?

To you question, currently this step runs when something commits into the raptor branch.

Comment (whimboo) [reply .circleci/config.yml]: On this repository we do not have any media files. Reason is that those are way too large, and also don't change (maybe at all).

If new tests are added which require new media files, those would have to be downloaded first, and only then this repository has to be checked for updates. So best would be to have it manually, or?

Also please note that the target website is used frequently by automated tests. So updating the media files directly at the target location could cause test failures in Taskcluster. Maybe media files should first be downloaded to a separate folder, and then replaced?

**Comment by whimboo** [comment]
Please note that CircleCI also failed to parse the YAML file.

File: .circleci/config.yml
Comment (bqbn) [reply .circleci/config.yml]: Ok, that sounds good. We'll update the media files manually whenever it's needed.

I updated this PR so that CircleCI only deploy "code" not "media".

Since updating the media files is going to be very rare, I think it can be easily coordinated when to run it. Let me know it must be an atom operation though.

**Comment by bqbn** [comment]
> Please note that CircleCI also failed to parse the YAML file.

The syntax error is fixed.

File: .utils/deploy.sh
Comment (whimboo) [reply .utils/deploy.sh]: I'll leave this conversation open until it got a proper review from @g-k or any other team member.

File: .circleci/config.yml
Comment (whimboo) [reply .circleci/config.yml]: Should we create some documentation for how to update the media files?

Comment (bqbn) [reply .circleci/config.yml]: Sure thing. I've updated the PR with a README file.

File: .utils/deploy.sh
Comment (g-k) [inline .utils/deploy.sh]: This is great! If the static content isn't framed anywhere I'd add `frame-ancestors: 'none'` too since it doesn't fallback to `default-src`. 

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors

File: .utils/deploy.sh
Comment (g-k) [inline .utils/deploy.sh]: To put it mildly this is not so great. We're basically getting no XSS or clickjacking protection here. 

What is this CSP policy being used for? We should be able to make it less permissive.

File: .utils/deploy.sh
Comment (g-k) [inline .utils/deploy.sh:20]: :+1: for the above

File: .utils/deploy.sh
Comment (g-k) [reply .utils/deploy.sh]: Oh also if we have a version of Sentry with Fx CSP report support we could add a report-uri pointing there.

**Review by g-k — CHANGES_REQUESTED** [review]

File: .circleci/config.yml
Comment (whimboo) [reply .circleci/config.yml]: I would suggest that we also add this to the wiki of this repository, and maybe add a link inside the README. Please see https://github.com/mozilla/perf-youtube-playback/wiki

File: .utils/deploy.sh
Comment (bqbn) [reply .utils/deploy.sh]: @whimboo does your team have any requirement for the CSP header for this web site?

@g-k also, what would secops team want us to use?

From ops perspective, we can use whatever CSP headers you guys want. The current CSP header mimics what https://ytlr-cert.appspot.com/2019/main.html uses, and this is meant to be reviewed, since we (ops) don't know what the requirement is for this site, as far as CSP header is concerned.

File: .utils/deploy.sh
Comment (bqbn) [reply .utils/deploy.sh]: `frame-ancestors: 'none'` added.

> if we have a version of Sentry with Fx CSP report support we could add a report-uri pointing there.

We don't have that yet.

File: .utils/deploy.sh
Comment (whimboo) [reply .utils/deploy.sh]: I actually don't really know about the CSP headers. So I would defer to @g-k here. As long as the tests are working it's all fine.

**Comment by whimboo** [comment]
@bqbn do we have an update here? It would be great to get it landed.

**Comment by bqbn** [comment]
@g-k any comments regarding https://github.com/mozilla/perf-youtube-playback/pull/2#discussion_r292653683

Both ops and @whimboo can defer to secops about the CSP header values. Per @whimboo 

> As long as the tests are working it's all fine.

I think it's best for secops to propose some initial values, it they work for the tests, great. If not, we can iterate on those.

Thanks.

**Comment by g-k** [comment]
@bqbn my bad, I've been meaning to look into this. Can we try the following?

```
default-src 'none'; img-src 'self' data:; script-src 'self' 'unsafe-inline'; font-src 'self'; frame-ancestors 'none'; base-uri 'none'; style-src 'self' 'unsafe-inline'; connect-src 'self'
```

The main changes are getting rid of the `unsafe-eval` and `*` `script-src`s and moving those sources to the resource specific directives instead of `default-src`. It worked well for me in a meta element (loading the page and clicking a few links to load a video but there might be other functionality that CSP would block that I didn't test).

Who are the consumers of the site? If this is just for devs and QA and not for wider public consumption I won't worry about pulling the inline scripts into a separate file.

**Comment by whimboo** [comment]
> Who are the consumers of the site? If this is just for devs and QA and not for wider public consumption I won't worry about pulling the inline scripts into a separate file.

Yes, this is just for Raptor tests as run in CI automation, or locally for ensuring the test is still working fine.

**Comment by bqbn** [comment]
I've updated the CSP header according to the recommendation. Please review and approve accordingly.

Thanks.

**Review by g-k — APPROVED** [review]
r+ 

I was on PTO lgtm.

**Comment by whimboo** [comment]
Thanks a lot! I'm going to merge it now. @bqbn feel free to continue on the server side.

**Comment by bqbn** [comment]
@whimboo remind me if you can, but I can't think of anything needs to be done on the server side at this point.

Basically, now that this is merged, every time you push a commit to the "raptor" branch, the branch will be built and pushed to the environments.

Let me know if you have questions. Thanks!

**Comment by whimboo** [comment]
@bqbn please follow-up on the needinfo request on the appropriate bug in Bugzilla. Close it if nothing else needs to be done. Thanks!

**Comment by bqbn** [comment]
@whimboo I've closed the bugzilla ticket (https://bugzilla.mozilla.org/show_bug.cgi?id=1548062) as nothing that I can think of that needs to be done in that ticket.

However, there is something you need to do in this repo, which is to merge master (or cherry-pick the commits in this PR) into `rapter` branch, because I think that branch currently doesn't contain the `.circleci` directory, which means it won't build even if it's updated.

Thanks!

**Pull request merged**
